File: 05_prefix_sums/genomic_range_query.py
def solution(s, p, q):
  m = len(p)
  a = prefix_count(s, 'A')
  c = prefix_count(s, 'C')
  g = prefix_count(s, 'G')
  ret = []
  for i in xrange(m):
    start = p[i]
    end = q[i] + 1
    if a[end] - a[start] > 0:
      ret.append(1)
    elif c[end] - c[start] > 0:
      ret.append(2)
    elif g[end] - g[start] > 0:
      ret.append(3)
    else:
      ret.append(4)
  return ret

def prefix_count(s, char):
  n = len(s)
  ret = [0] * (n + 1)
  cnt = 0
  for i in xrange(n):
    if s[i] == char:
      cnt += 1
    ret[i+1] = cnt
  return ret

File: 05_prefix_sums/genomic_range_query.cpp
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

const int MAX = 100010;

vector<int> solution(string &s, vector<int> &p, vector<int> &q) {
  int n = s.length();
  int m = p.size();
  int cnts[4][MAX] = {{0}};
  string criteria = ("ACGT");
  vector<int> ret(m, 0);

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < 4; j++){
      cnts[j][i+1] = cnts[j][i];
      if (s[i] == criteria[j])
        cnts[j][i+1]++;
    }
  }
  
  for (int i = 0; i < m; i++) {
    for (int j = 0; j < 4; j++){
      if (cnts[j][q[i] + 1] - cnts[j][p[i]] > 0){
        ret[i] = j + 1;
        break;
      }
    }
  }
  return ret;
}
